### PR TITLE
get encoder library from smaller repo, avoid silently failing if enco…

### LINKED ISSUE #129: qsc_encodings silently failing
**Describe the bug**
The CI tests show error messages like "Could not create OQS signature encoding algorithm draft-uni-qsckeys-falcon-00/sk-pk (Falcon-512, 1.3.9999.3.6). Defaulting to no encoding."

**To Reproduce**
See for example https://app.circleci.com/pipelines/github/open-quantum-safe/oqs-provider/410/workflows/a97aa465-7023-4479-89e0-844fcb926e7f/jobs/431 (scroll to last test sequence and check logfiles).

**Expected behavior**
The tests should fail and not pass -- otherwise they don't have value, right @bhess?

**Environment (please complete the following information):**
CircleCI

Update/add: I cannot reproduce this behaviour locally, so here's at least a screenshot:
grafik

### LINKED ISSUE #130: Improve qsc encoder integration
Build time for oqs-provider significantly deteriorated when qsc-encoder integration was done. A glance at disk usage shows the problem vividly: The `src` of qsc-`encoder` is more than 2x as large as all of OpenSSL, 3x the size of a fully built `liboqs` and thus roughly 60% of a complete `oqs-provider` build (incl. opensslv3 binaries and libraries):

grafik

--> Is it really necessary to add KATs for all algorithms (again)? That alone seems to be 255MBytes. Then there's PPTX files and huge `.git` index files: The latter are 38x larger in qsc-encoder than in all of oqs-provider (3.5MB vs 138MB).

@bhess: Do you agree that this is not OK and have time to remedy this?

### PR DESCRIPTION
**Pull request by bhess**
- Gets encoder library from smaller repository
- Avoid encoding tests silently fail if algorithm/encoding is not found
- Points to new upstream encoder version with updated OID
- Adds `UPDATE_DISCONNECTED 1` to allow offline build (#139)

Fixes #129 
Fixes #130

### CODE REVIEW

File: oqsprov/oqsprov_keys.c [hunk 1]
```diff
@@ -619,9 +619,10 @@ OQSX_KEY *oqsx_key_new(OSSL_LIB_CTX *libctx, char* oqs_name, char* tls_name, int
 #ifdef USE_ENCODING_LIB
         if (alg_idx >= 0 && oqs_alg_encoding_list[alg_idx] != NULL) {
             if (qsc_encoding_by_name_oid(&ret->oqsx_encoding_ctx.encoding_ctx, &ret->oqsx_encoding_ctx.encoding_impl, oqs_oid_alg_list[2*alg_idx], oqs_alg_encoding_list[alg_idx]) != QSC_ENC_OK) {
-                fprintf(stderr, "Could not create OQS signature encoding algorithm %s (%s, %s). Defaulting to no encoding.\n", oqs_alg_encoding_list[alg_idx], oqs_name, oqs_oid_alg_list[2*alg_idx]);
-            ret->oqsx_encoding_ctx.encoding_ctx = NULL;
+                fprintf(stderr, "Could not create OQS signature encoding algorithm %s (%s, %s).\n", oqs_alg_encoding_list[alg_idx], oqs_name, oqs_oid_alg_list[2*alg_idx]);
```
Comment (baentsch) [inline oqsprov/oqsprov_keys.c:622]: OK -- you don't output that you're not defaulting to no encoding. But the code doesn't change -- so still defaults on error, correct?

**Review by baentsch — CHANGES_REQUESTED** [review]
Thanks for the fixes. But I fail to see a behavioural change in the code, so am I assuming right that it will fail in the same way if the OIDs between qsc-encoding lib and oqs-provider go out of sync again? Also, wouldn't it be more professional to have tests for this (both good and negative expected behaviour)? Finally, is such duplication of OIDs between oqs-provider and qsc-encoder really necessary (or at the very least cannot be overruled at the API level)? It creates an undue dependency that we've just fallen into -- and can fall into again if I read https://github.com/Quantum-Safe-Collaboration/qsc-key-encoder/commit/d7970eaa874182aec042e474b7c78316c82332d2 correctly.

It is understood this is a quick fix to help Goutam this weekend, but I'd like to bring this PR to a proper completion.

@xvzcf: Please take this branch as a hotfix if you can still need it in tomorrows testing session. I also published a docker image at openquantumsafe/oqs-ossl3:ietf116-hotfix[-dev] to make your life easier.

File: oqsprov/oqsprov_keys.c (hunk 1 above)
Comment (bhess) [reply oqsprov/oqsprov_keys.c:622]: it adds a `goto err`, which will raise an error.

**Comment by bhess** [comment]
> Thanks for the fixes. But I fail to see a behavioural change in the code, so am I assuming right that it will fail in the same way if the OIDs between qsc-encoding lib and oqs-provider go out of sync again? Also, wouldn't it be more professional to have tests for this (both good and negative expected behaviour)? 

It will not silently fail anymore. Good point with the test, I can add a negative test (show that the tests fail if a unknown encoding is set via env var).

> Finally, is such duplication of OIDs between oqs-provider and qsc-encoder really necessary (or at the very least cannot be overruled at the API level)? It creates an undue dependency that we've just fallen into -- and can fall into again if I read [Quantum-Safe-Collaboration/qsc-key-encoder@d7970ea](https://github.com/Quantum-Safe-Collaboration/qsc-key-encoder/commit/d7970eaa874182aec042e474b7c78316c82332d2) correctly.

I'm not sure what's the best solution. How would you overrule at API level?
There needs to be a unique key to match. Another way the API supports is lookup by algorithm name. Here the issue is that the names are not really well defined, e.g, OQS name: `SPHINCS+-SHAKE256-128s-robust` and the name used in the reference implementation: `sphincs-shake256-128s-robust`. The names also don't identify the algorithm version. OIDs seem to define things more uniquely (ideally there will be a OID repo to pull from). The Falcon OID change was unrelated to key encoding (but signatures changed). Next time the key encoding could change which would require a new IETF draft revision.

**Comment by baentsch** [comment]
> How would you overrule at API level? 

Client (`oqs-provider`) requests algorithm name (of qsc-encoder) w/o OID reference: It should be the responsibility of the client that algorithm name and OID match (or not, e.g., at an interop hackathon where people play with OIDs).

> There needs to be a unique key to match. 

Only when there are different encodings for the same algorithm name, right? Are there already different encoding for the same algorithm (name)? If it were coming to that, what about a SHA(256) over an (agreed upon) KAT to identify algs with the same name (but different versions/KATs)?

File: oqsprov/oqsprov_keys.c (hunk 1 above)
Comment (baentsch) [reply oqsprov/oqsprov_keys.c:622]: Oops -- didn't see that. Thanks.

**Review by baentsch — COMMENTED** [review]
Thanks for explanations and update. I presume testing (good/bad cases) and documentation for these changes (env vars) will be added in further commits to this PR (?)

**Comment by baentsch** [comment]
> Added a commit with both.

Thanks. Then what am I doing wrong here (after the tests ran, using the resultant .crt files):

As expected:
```
> OPENSSL_MODULES=_build/lib LD_LIBRARY_PATH=.local/lib64 .local/bin/openssl x509 -in tmp/falcon1024.crt -text -noout -provider oqsprovider -provider default
[...]
        Subject: CN=test Server
        Subject Public Key Info:
            Public Key Algorithm: falcon1024
                falcon1024 public key:
                PQ key material:
[...]
```

Not quite expected:
```
> OPENSSL_MODULES=_build/lib LD_LIBRARY_PATH=.local/lib64 .local/bin/openssl x509 -in tmp/falcon512.crt -text -noout -provider oqsprovider -provider default
[...]
        Subject: CN=test Server
        Subject Public Key Info:
            Public Key Algorithm: falcon512
            Unable to load Public Key
808B1AEC977F0000:error:03000072:digital envelope routines:X509_PUBKEY_get0:decode error:crypto/x509/x_pubkey.c:463:
808B1AEC977F0000:error:03000072:digital envelope routines:X509_PUBKEY_get0:decode error:crypto/x509/x_pubkey.c:463:
        X509v3 extensions:
[...]
```

Trying with 
```
OPENSSL_MODULES=_build/lib OQS_ENCODING_FALCON512=draft-uni-qsckeys-falcon-00/sk-pk OQS_ENCODING_FALCON512_ALGNAME=Falcon512 LD_LIBRARY_PATH=.local/lib64 .local/bin/openssl x509 -in tmp/falcon512.crt -text -noout -provider oqsprovider -provider default
```
yields the same result.

What did I do? Changed the Falcon512 OID. --> Is there still some dependency on it?

**Comment by bhess** [comment]
I could reproduce the error, however, I get an error even without changing the OID (Falcon1024 looks fine, Falcon512 shows the error). I will need to investigate why. There shouldn't be any OID dependency. 

> Not quite expected:
> 
> ```
> > OPENSSL_MODULES=_build/lib LD_LIBRARY_PATH=.local/lib64 .local/bin/openssl x509 -in tmp/falcon512.crt -text -noout -provider oqsprovider -provider default
> [...]
>         Subject: CN=test Server
>         Subject Public Key Info:
>             Public Key Algorithm: falcon512
>             Unable to load Public Key
> 808B1AEC977F0000:error:03000072:digital envelope routines:X509_PUBKEY_get0:decode error:crypto/x509/x_pubkey.c:463:
> 808B1AEC977F0000:error:03000072:digital envelope routines:X509_PUBKEY_get0:decode error:crypto/x509/x_pubkey.c:463:
>         X509v3 extensions:
> [...]
> ```

Is this really the ouput after running the tests with changed Falcon512 OID (via env var)?

I tried it as well - first without using the external encoder, changing the Falcon512 OID via env var. Even here I get an error, but a slightly different one:

```
Subject: CN=test Server
        Subject Public Key Info:
            Public Key Algorithm: 1.3.9999.3.6.10
            Unable to load Public Key
4001FBEF01000000:error:03000072:digital envelope routines:X509_PUBKEY_get0:decode error:crypto/x509/x_pubkey.c:463:
4001FBEF01000000:error:03000072:digital envelope routines:X509_PUBKEY_get0:decode error:crypto/x509/x_pubkey.c:463:
        X509v3 extensions:
```

The "Public Key Algorithm" seems to have changed from 'falcon512' to the OID I set via env var. Shouldn't the name be unchanged?

Perhaps there are two different (maybe related) issues?
(1) when changing the OIDs -> applies not only when using the external encoder, and 
(2) the error when parsing the falcon512, only happens when using the external encoder

**Comment by bhess** [comment]
> I could reproduce the error, however, I get an error even without changing the OID (Falcon1024 looks fine, Falcon512 shows the error). I will need to investigate why. There shouldn't be any OID dependency.

I didn't set all env variables correctly when testing this. Trying again, it seems to all work ok, including changing the OID.

Testing Falcon with changed OIDs:

```
export OQS_ENCODING_FALCON512=draft-uni-qsckeys-falcon-00/sk-pk
export OQS_ENCODING_FALCON1024=draft-uni-qsckeys-falcon-00/sk-pk
export OQS_ENCODING_FALCON512_ALGNAME=Falcon512
export OQS_ENCODING_FALCON1024_ALGNAME=Falcon1024
export OQS_OID_FALCON512=1.3.9999.3.20
export OQS_OID_FALCON1024=1.3.9999.3.21
```

- run `scripts/runtests_encodings.sh`

- test Falcon1024
```
$ OPENSSL_MODULES=_build/lib LD_LIBRARY_PATH=.local/lib64 .local/bin/openssl x509 -in tmp/falcon1024.crt -text -noout -provider oqsprovider -provider default
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            52:1a:85:d8:b7:66:c8:0f:f6:86:07:4a:eb:24:6c:c9:55:94:8b:7d
        Signature Algorithm: falcon1024
        Issuer: CN=test CA
        Validity
            Not Before: Jan 23 08:00:00 2015 GMT
            Not After : Aug 23 09:00:00 2025 GMT
        Subject: CN=test Server
        Subject Public Key Info:
            Public Key Algorithm: falcon1024
                falcon1024 public key:
                PQ key material:
                    0a:70:9a:b5:96:8c:e7:e3:80:ae:9b:80:3c:21:c0:
```

- test Falcon512
```
% OPENSSL_MODULES=_build/lib LD_LIBRARY_PATH=.local/lib64 .local/bin/openssl x509 -in tmp/falcon512.crt -text -noout -provider oqsprovider -provider default 
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            13:b7:60:75:a4:0e:99:d8:a9:6f:07:61:5d:7d:e1:aa:05:9a:1f:15
        Signature Algorithm: falcon512
        Issuer: CN=test CA
        Validity
            Not Before: Jan 23 08:00:00 2015 GMT
            Not After : Aug 23 09:00:00 2025 GMT
        Subject: CN=test Server
        Subject Public Key Info:
            Public Key Algorithm: falcon512
                falcon512 public key:
                PQ key material:
                    09:b4:d2:f4:28:df:03:a1:42:3d:89:c9:a4:26:4c:
```

After `unset OQS_OID_FALCON512`, the above test fails as expected with a decoding error.


> What did I do? Changed the Falcon512 OID. --> Is there still some dependency on it?

@baentsch looking at your test command a `OQS_OID_FALCON512=...` might be missing.

**Comment by baentsch** [comment]
> @baentsch looking at your test command a OQS_OID_FALCON512=... might be missing.

Nope. I changed `generate.yml` for the test...

**Comment by baentsch** [comment]
Thanks for these fixes. I hope you can forgive me now being a "bit nervous" about the robustness of all this -- my question thus: Does https://github.com/IETF-Hackathon/pqc-certificates/pull/43 contain proper (proper in the sense of "in line with your RFC") certificates? Or should this data set be re-done after landing this PR to permit others to test "the right stuff"? Last question: Are you going to rebase the PR to remove the conflict?

**Comment by bhess** [comment]
> Thanks for these fixes. I hope you can forgive me now being a "bit nervous" about the robustness of all this -- my question thus: Does [IETF-Hackathon/pqc-certificates#43](https://github.com/IETF-Hackathon/pqc-certificates/pull/43) contain proper (proper in the sense of "in line with your RFC") certificates? Or should this data set be re-done after landing this PR to permit others to test "the right stuff"?

The certs produced with the "hotfix" docker image are equivalent with the ones now in this PR. Thanks for being extra cautious! 

> Last question: Are you going to rebase the PR to remove the conflict?

Done.

**Comment by baentsch** [comment]
> The certs produced with the "hotfix" docker image are equivalent with the ones now in this PR.

Thanks for checking (and re-basing).

**Review by baentsch — APPROVED** [review]
Thanks for the various fixes, @bhess: All works as desired. And in one instance better than I thought possible: After changing an OID via env var during cert gen, it is even possible to read certs generated that way without explicitly setting the changed OID as env var also to the `openssl x509 -text` command. Hmm. Would you know why this works? Some redundancy in the SPKI info?
